mypy/ipc.py

```diff
@@ -0,0 +1,185 @@
+        if sys.platform == 'win32':
+            try:
+                _winapi.WriteFile(self.connection, data)
+                # this empty write is to copy the behavior of socket.sendall,
+                # which also sends an empty message to signify it is done writing
+                _winapi.WriteFile(self.connection, b'')
+            except WindowsError as e:
+                raise IPCException("Failed to write with error: {}".format(e.winerror))
+        else:
+            self.connection.sendall(data)
+            self.connection.shutdown(socket.SHUT_WR)
+
+    def close(self) -> None:
+        if sys.platform == 'win32':
+            if self.connection != _winapi.NULL:
+                _winapi.CloseHandle(self.connection)
+        else:
+            self.connection.close()
+
+    def __exit__(self,
+                 exc_ty: 'Optional[Type[BaseException]]' = None,
+                 exc_val: Optional[BaseException] = None,
+                 exc_tb: Optional[TracebackType] = None,
+                 ) -> bool:
+        self.close()
+        return False
+
+
+class IPCClient(IPCBase):
+    """The client side of an IPC connection."""
+
+    def __init__(self, name: str, timeout: Optional[int]) -> None:
+        super().__init__(name)
+        if sys.platform == 'win32':
+            timeout = timeout or 1000  # we need to set a timeout
+            while True:
+                try:
+                    _winapi.WaitNamedPipe(self.name, timeout)
+                except WindowsError as e:
+                    if e.winerror == _winapi.ERROR_SEM_TIMEOUT:
+                        raise IPCException("Timed out waiting for connection.")
+                try:
+                    self.connection = _winapi.CreateFile(
+                        self.name,
+                        _winapi.GENERIC_READ | _winapi.GENERIC_WRITE,
+                        0,
+                        _winapi.NULL,
+                        _winapi.OPEN_EXISTING,
+                        0,
+                        _winapi.NULL,
+                    )
+                except WindowsError as e:
+                    if e.winerror == _winapi.ERROR_PIPE_BUSY:
+                        raise IPCException("The connection is busy.")
+                else:
+                    break
+            _winapi.SetNamedPipeHandleState(self.connection,
+                                            _winapi.PIPE_READMODE_MESSAGE,
+                                            None,
+                                            None)
+        else:
+            self.connection = socket.socket(socket.AF_UNIX)
+            self.connection.settimeout(timeout)
+            self.connection.connect(name)
+
+    def __enter__(self) -> 'IPCClient':
+        return self
+
+
+class IPCServer(IPCBase):
+
+    BUFFER_SIZE = 2**16
+
+    def __init__(self, name: str, timeout: Optional[int] = None) -> None:
+        super().__init__(name)
+        if sys.platform != 'win32':
```

Comment:
Let's unify the interfaces by making the server always be responsible for generating a unique socket name. On windows we can probably just grab some bytes from `os.urandom()` and call it a day.